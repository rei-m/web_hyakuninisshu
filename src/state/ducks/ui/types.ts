import { Action } from 'redux';
import { Color, Kimariji } from '@src/types';
import * as constants from './constants';

export type OPEN_KARUTAS_FILTER_TYPE = typeof constants.OPEN_KARUTAS_FILTER_NAME;

export type CLOSE_KARUTAS_FILTER_TYPE = typeof constants.CLOSE_KARUTAS_FILTER_NAME;

export type TOGGLE_KARUTAS_COLOR_TYPE = typeof constants.TOGGLE_KARUTAS_COLOR_NAME;

export type TOGGLE_KARUTAS_KIMARIJI_TYPE = typeof constants.TOGGLE_KARUTAS_KIMARIJI_NAME;

export interface OpenKarutasFilterAction extends Action {
  readonly type: OPEN_KARUTAS_FILTER_TYPE;
}

export interface CloseKarutasFilterAction extends Action {
  readonly type: CLOSE_KARUTAS_FILTER_TYPE;
}

export interface OKarutasFilterAction extends Action {
  readonly type: OPEN_KARUTAS_FILTER_TYPE;
}

export interface ToggleKarutasColorAction extends Action {
  readonly type: TOGGLE_KARUTAS_COLOR_TYPE;
  readonly payload: {
    readonly color: Color;
    readonly checked: boolean;
  };
}

export interface ToggleKarutasKimarijiAction extends Action {
  readonly type: TOGGLE_KARUTAS_KIMARIJI_TYPE;
  readonly payload: {
    readonly kimariji: Kimariji;
    readonly checked: boolean;
  };
}

export type Actions =
  | OpenKarutasFilterAction
  | CloseKarutasFilterAction
  | ToggleKarutasColorAction
  | ToggleKarutasKimarijiAction;

export interface KarutasFilter {
  readonly colors: Array<{ color: Color; checked: boolean }>;
  readonly kimarijis: Array<{ kimariji: Kimariji; checked: boolean }>;
}

export interface UiState {
  readonly karutasFilter: KarutasFilter & { readonly open: boolean };
}
