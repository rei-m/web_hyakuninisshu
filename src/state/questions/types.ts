import { Action } from 'redux';
import { Kimariji, Color, QuestionId, KarutaNo, Karuta } from '@src/domain/models';
import * as constants from './constants';
import { Meta, Payload } from '..';

type Condition<T> = {
  values: Array<T>;
};

export type RangeFromCondition = 1 | 11 | 21 | 31 | 41 | 51 | 61 | 71 | 81 | 91;
export const RangeFromCondition: Condition<RangeFromCondition> = {
  values: [1, 11, 21, 31, 41, 51, 61, 71, 81, 91],
};

export type RangeToCondition = 10 | 20 | 30 | 40 | 50 | 60 | 70 | 80 | 90 | 100;
export const RangeToCondition: Condition<RangeToCondition> = {
  values: [10, 20, 30, 40, 50, 60, 70, 80, 90, 100],
};

export type KimarijiCondition = null | Kimariji;
export const KimarijiCondition: Condition<KimarijiCondition> = {
  values: [null, 1, 2, 3, 4, 5, 6],
};

export type ColorCondition = null | Color;
export const ColorCondition: Condition<ColorCondition> = {
  values: [null, 'blue', 'pink', 'yellow', 'green', 'orange'],
};

export type KarutaStyleCondition = 'kanji' | 'kana';
export const KarutaStyleCondition: Condition<KarutaStyleCondition> = {
  values: ['kanji', 'kana'],
};

export type QuestionAnimCondition = 'none' | 'slow' | 'normal' | 'fast';
export const QuestionAnimCondition: Condition<QuestionAnimCondition> & {
  toDulation: (value: QuestionAnimCondition) => number;
} = {
  values: ['none', 'slow', 'normal', 'fast'],
  toDulation: (value: QuestionAnimCondition) => {
    switch (value) {
      case 'normal':
        return 0.6;
      case 'fast':
        return 0.3;
      case 'slow':
        return 1.0;
      default:
        return 0;
    }
  },
};

export type YomiFuda = Readonly<{
  karutaNo: KarutaNo;
  shoku: string;
  niku: string;
  sanku: string;
}>;

export type ToriFuda = Readonly<{
  karutaNo: KarutaNo;
  shiku: string;
  kekku: string;
}>;

export type StartTrainingAction = Action<typeof constants.START_TRAINING_NAME> &
  Payload<{
    currentQuestionId: QuestionId;
    totalCount: number;
  }> &
  Meta<{
    rangeFrom: RangeFromCondition;
    rangeTo: RangeToCondition;
    kimariji: KimarijiCondition;
    color: ColorCondition;
    kamiNoKuStyle: KarutaStyleCondition;
    shimoNoKuStyle: KarutaStyleCondition;
    questionAnim: QuestionAnimCondition;
  }>;

export type RestartTrainingAction = Action<typeof constants.RESTART_TRAINING_NAME> &
  Payload<{
    currentQuestionId: QuestionId;
    totalCount: number;
  }>;

export type StartExamAction = Action<typeof constants.START_EXAM_NAME> &
  Payload<{
    currentQuestionId: QuestionId;
    totalCount: number;
  }>;

export type StartQuestionAction = Action<typeof constants.START_QUESTION_NAME> &
  Payload<{
    questionId: QuestionId;
    content: {
      yomiFuda: YomiFuda;
      toriFudaList: [ToriFuda, ToriFuda, ToriFuda, ToriFuda];
    };
  }>;

export type ResetQuestionAction = Action<typeof constants.RESET_QUESTION_NAME>;

export type FinishQuestionAction = Action<typeof constants.FINISH_QUESTION_NAME> &
  Payload<{
    correctCount: number;
    averageAnswerSecond: number;
    answerList: Array<{
      questionId: QuestionId;
      isCorrect: boolean;
      correctKaruta: Karuta;
    }>;
  }>;

export type AnswerQuestionAction = Action<typeof constants.ANSWER_QUESTION_NAME> &
  Payload<{
    isCorrect: boolean;
    selectedKarutaNo: KarutaNo;
    correctKaruta: Karuta;
  }>;

export type ConfirmCorrectAction = Action<typeof constants.CONFIRM_CORRECT_NAME>;

export type OpenNextQuestionAction = Action<typeof constants.OPEN_NEXT_QUESTION_NAME> &
  Payload<{
    currentQuestionId: QuestionId;
  }>;

export type Actions =
  | StartTrainingAction
  | RestartTrainingAction
  | StartExamAction
  | StartQuestionAction
  | ResetQuestionAction
  | FinishQuestionAction
  | AnswerQuestionAction
  | ConfirmCorrectAction
  | OpenNextQuestionAction;

export type ActionCreator = {
  /**
   * 指定した条件から問題を作成/保存してStartTrainingActionを返す
   *
   * @param rangeFrom 出題範囲From
   * @param rangeTo 出題範囲To
   * @param kimariji 決まり字条件
   * @param color 色条件
   * @param kamiNoKuStyle 上の句表示形式
   * @param shimoNoKuStyle 下の句表示形式
   * @param questionAnim 問題の表示速度
   *
   * @returns {StartTrainingAction}
   * @throws {Error} 指定した条件に当てはまる歌が無い場合
   */
  startTraining: (
    rangeFrom: RangeFromCondition,
    rangeTo: RangeToCondition,
    kimariji: KimarijiCondition,
    color: ColorCondition,
    kamiNoKuStyle: KarutaStyleCondition,
    shimoNoKuStyle: KarutaStyleCondition,
    questionAnim: QuestionAnimCondition
  ) => StartTrainingAction;

  /**
   * 回答済の問題の中から間違えた問題をもとに問題を作成/保存してRestartTrainingActionを返す
   *
   * @returns {RestartTrainingAction}
   * @throws {Error} 指定した条件に当てはまる歌が無い場合
   */
  restartTraining: () => RestartTrainingAction;

  /**
   * 全ての歌を対象に問題を作成/保存してStartExamActionを返す
   *
   * @returns {StartExamAction}
   * @throws {Error} 指定した条件に当てはまる歌が無い場合
   */
  startExam: () => StartExamAction;

  /**
   * 問題の回答を開始してStartQuestionActionを返す
   *
   * @returns {StartQuestionAction}
   * @throws {Error} 指定した問題がない場合
   */
  startQuestion: (
    currentQuestionId: QuestionId,
    kamiNoKuStyle: KarutaStyleCondition,
    shimoNoKuStyle: KarutaStyleCondition,
    startDate: Date
  ) => StartQuestionAction;

  /**
   * 問題の回答を判定してAnswerQuestionActionを返す
   *
   * @returns {AnswerQuestionAction}
   * @throws {Error} 指定した問題がない場合
   * @throws {Error} 問題の回答が開始されていない場合
   */
  answerQuestion: (currentQuestionId: QuestionId, toriFuda: ToriFuda, answerDate: Date) => AnswerQuestionAction;

  /**
   * ConfirmCorrectActionを返す
   *
   * @returns {ConfirmCorrectAction}
   */
  confirmCorrect: () => ConfirmCorrectAction;

  /**
   * 次の問題を取得してOpenNextQuestionActionを返す
   *
   * @returns {OpenNextQuestionAction}
   * @throws {Error} 次の問題がない場合
   */
  openNextQuestionAction: (currentQuestionId: QuestionId) => OpenNextQuestionAction;

  /**
   * 問題を初期状態に戻してResetQuestionActionを返す
   *
   * @returns {ResetQuestionAction}
   */
  resetQuestion: () => ResetQuestionAction;

  /**
   * 問題を終了して結果を見る
   *
   * @returns {FinishQuestionAction}
   */
  finishQuestion: () => FinishQuestionAction;
};

export type State = Readonly<{
  state: 'waiting' | 'ready' | 'playing' | 'confirm' | 'finished';
  totalCount: number;
  currentPosition: number;
  currentQuestion?: {
    questionId: QuestionId;
    content?: {
      yomiFuda: YomiFuda;
      toriFudaList: [ToriFuda, ToriFuda, ToriFuda, ToriFuda];
    };
    answer?: {
      isCorrect: boolean;
      selectedKarutaNo: KarutaNo;
      correctKaruta: Karuta;
    };
  };
  result?: {
    correctCount: number;
    averageAnswerSecond: number;
    answerList: Array<{
      questionId: QuestionId;
      isCorrect: boolean;
      correctKaruta: Karuta;
    }>;
  };
  trainingCondition: {
    rangeFrom: RangeFromCondition;
    rangeTo: RangeToCondition;
    kimariji: KimarijiCondition;
    color: ColorCondition;
    kamiNoKuStyle: KarutaStyleCondition;
    shimoNoKuStyle: KarutaStyleCondition;
    questionAnim: QuestionAnimCondition;
  };
}>;
