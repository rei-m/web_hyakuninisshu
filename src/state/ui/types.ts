import { Action } from 'redux';
import * as constants from './constants';
import { Color, Kimariji } from '@src/domain/models';
import { Payload } from '..';

export type KarutasFilter = Readonly<{
  colors: Array<{ color: Color; checked: boolean }>;
  kimarijis: Array<{ kimariji: Kimariji; checked: boolean }>;
}>;

export type OpenKarutasFilterAction = Action<typeof constants.OPEN_KARUTAS_FILTER_NAME>;

export type CloseKarutasFilterAction = Action<typeof constants.CLOSE_KARUTAS_FILTER_NAME>;

export type ToggleKarutasColorAction = Action<typeof constants.TOGGLE_KARUTAS_COLOR_NAME> &
  Payload<{
    color: Color;
    checked: boolean;
  }>;

export type ToggleKarutasKimarijiAction = Action<typeof constants.TOGGLE_KARUTAS_KIMARIJI_NAME> &
  Payload<{
    kimariji: Kimariji;
    checked: boolean;
  }>;

export type Actions =
  | OpenKarutasFilterAction
  | CloseKarutasFilterAction
  | ToggleKarutasColorAction
  | ToggleKarutasKimarijiAction;

export type ActionCreator = {
  /**
   * OpenKarutasFilterActionを返す
   *
   * @returns {OpenKarutasFilterAction}
   */
  openKarutasFilter: () => OpenKarutasFilterAction;

  /**
   * CloseKarutasFilterActionを返す
   *
   * @returns {CloseKarutasFilterAction}
   */
  closeKarutasFilter: () => CloseKarutasFilterAction;

  /**
   * ToggleKarutasColorActionを返す
   *
   * @param color 歌の色
   * @param checked 表示する/しない
   *
   * @returns {ToggleKarutasColorAction}
   */
  toggleKarutasColor: (color: Color, checked: boolean) => ToggleKarutasColorAction;

  /**
   * ToggleKarutasKimarijiActionを返す
   *
   * @param kimariji 歌の決まり字
   * @param checked 表示する/しない
   *
   * @returns {ToggleKarutasKimarijiAction}
   */
  toggleKarutasKimariji: (kimariji: Kimariji, checked: boolean) => ToggleKarutasKimarijiAction;
};

export type State = Readonly<{
  karutasFilter: KarutasFilter & { open: boolean };
}>;
