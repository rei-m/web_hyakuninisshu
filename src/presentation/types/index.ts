import { RouteComponentProps } from '@reach/router';

export type SiteMetaData = {
  title: string;
  description: string;
  author: string;
};

export type GeneratedPageComponentProps<T = {}> = { pageContext: T } & RouteComponentProps<T>;

export type AppErrorType = 'Network' | 'NotFound' | 'Unknown';

export class AppError implements Error {
  private _message: string;
  private _type: AppErrorType;

  constructor(message: string, type: AppErrorType) {
    this._message = message;
    this._type = type;
  }

  public get name() {
    return 'AppError';
  }

  public get message() {
    return this._message;
  }

  public get type() {
    return this._type;
  }

  public toString() {
    return `${this.name}: ${this.message}`;
  }
}
