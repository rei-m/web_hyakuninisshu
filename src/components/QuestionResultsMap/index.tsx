import * as React from 'react';
import styled from 'styled-components';
import { withAppTheme } from '../../styles';
import { Answer, Karuta, Question } from '../../types';
import { toKarutaIdString } from '../helper';
import * as correctImage from './check_correct.png';
import * as incorrectImage from './check_incorrect.png';

export interface QuestionResultsMapProps {
  readonly questions: Question[];
  readonly answers: Answer[];
  readonly style?: React.CSSProperties;
  readonly onClickResult: (karuta: Karuta) => void;
}

export interface QuestionResultsCellProps {
  readonly question: Question;
  readonly answer: Answer;
  readonly onClickResult: (karuta: Karuta) => void;
}

const Root = withAppTheme(styled.div)`
  display: flex;
  flex-wrap: wrap;
  background-color: ${({ theme }) => theme.colorThin};
  box-sizing: border-box;
  box-shadow: ${({ theme }) => theme.elevationShadow1x};
  border-radius: 4px;
`;

const Cell = styled.div`
  width: 20%;
  height: 69px;
  position: relative;
  display: flex;
  align-items: center;
  justify-content: center;
`;

const KarutaNo = styled.span`
  font-size: 1.2rem;
`;

const CorrentImageBox = styled.span`
  position: absolute;
  top: 0;
  left: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 100%;
`;

export const QuestionResultsCell = ({
  question,
  answer,
  onClickResult
}: QuestionResultsCellProps) => {
  const onClickCell = () => {
    onClickResult(question.correctKaruta);
  };
  return (
    <Cell onClick={onClickCell}>
      <KarutaNo>{toKarutaIdString(question.correctKaruta.id)}</KarutaNo>
      <CorrentImageBox>
        <img
          src={answer.correct ? correctImage : incorrectImage}
          style={{ width: '80%' }}
        />
      </CorrentImageBox>
    </Cell>
  );
};

const QuestionResultsMap = ({
  questions,
  answers,
  style,
  onClickResult
}: QuestionResultsMapProps) => (
  <Root style={style}>
    {questions.map((q, i) => (
      <QuestionResultsCell
        question={q}
        answer={answers[i]}
        onClickResult={onClickResult}
        key={i}
      />
    ))}
  </Root>
);

export default QuestionResultsMap;
