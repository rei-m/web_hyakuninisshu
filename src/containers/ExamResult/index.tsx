import * as React from 'react';
import { connect } from 'react-redux';
import { navigate } from 'gatsby';
import ExamResultView from '@src/components/ExamResultView';
import ErrorMessage from '@src/components/ErrorMessage';
import { GlobalState } from '@src/state';
import { QuestionState } from '@src/enums';
import { Answer, Question } from '@src/types';
import { ROUTE_PATHS } from '@src/constants';

export interface ConnectedProps {
  questions: Question[];
  answers: Answer[];
  totalCount: number;
  questionState?: QuestionState;
}

export type Props = ConnectedProps;

export const ExamResult: React.FC<Props> = ({ questions, answers, totalCount, questionState }) => {
  if (questionState !== QuestionState.Finished) {
    return <ErrorMessage text="不正な遷移を行いました。前の画面からやり直してください。" />;
  }
  const averageAnswerSecond = answers.reduce((prev, current) => prev + current.time, 0) / 1000 / totalCount;

  const onClickRestartHandler = () => {
    navigate(ROUTE_PATHS.TRAINING_QUESTION, {
      state: {
        submitTime: new Date().getTime(),
        restart: true,
      },
    });
  };

  return (
    <ExamResultView
      averageAnswerSecond={Math.round(averageAnswerSecond * 100) / 100}
      totalCount={totalCount}
      correctCount={answers.filter(a => a.correct).length}
      answers={answers}
      questions={questions}
      onClickRestart={onClickRestartHandler}
    />
  );
};

export const mapStateToProps = ({ questions }: GlobalState): ConnectedProps => {
  const { answers, questionState } = questions;
  return {
    questions: questions.questions,
    answers,
    totalCount: questions.questions.length,
    questionState,
  };
};

export default connect(mapStateToProps)(ExamResult);
