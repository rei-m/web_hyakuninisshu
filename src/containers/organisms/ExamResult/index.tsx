import * as React from 'react';
import { connect } from 'react-redux';
import styled from '@src/styles/styled-components';
import Block from '@src/components/atoms/Block';
import ClosableDialog from '@src/components/molecules/ClosableDialog';
import QuestionJudgement from '@src/components/molecules/QuestionJudgement';
import KarutaPlayingResult from '@src/components/organisms/KarutaPlayingResult';
import Material from '@src/components/organisms/Material';
import { GlobalState } from '@src/state';
import { Answer, Karuta, Question } from '@src/types';

export interface OwnProps {
  onClickBack: () => void;
  onClickRestart: () => void;
}

export interface ConnectedProps {
  questions: Question[];
  answers: Answer[];
}

export type Props = OwnProps & ConnectedProps;

export type PresenterProps = Props & {
  displayedKaruta?: Karuta;
  onClickResult: (karuta: Karuta) => void;
  onCloseDialog: () => void;
};

export type ContainerProps = Props & {
  presenter: (props: PresenterProps) => React.ReactElement;
};

const useMaterialDialog = (): [Karuta | undefined, (karuta: Karuta) => void, () => void] => {
  const [karuta, setKaruta] = React.useState<Karuta>();
  const closeDialog = React.useCallback(() => {
    setKaruta(undefined);
  }, []);

  return [karuta, setKaruta, closeDialog];
};

const StyledBlock = styled(Block)`
  display: flex;
  flex-wrap: wrap;
  background-color: ${({ theme }) => theme.colorThin};
  box-sizing: border-box;
  box-shadow: ${({ theme }) => theme.elevationShadow1x};
  border-radius: 4px;
  margin: ${({ theme }) => theme.spacing2x} 0;
`;

const StyledQuestionJudgement = styled(QuestionJudgement)`
  width: 20%;
  height: 69px;
  padding: ${({ theme }) => theme.spacing2x};
`;

export const ExamResultPresenter = ({
  questions,
  answers,
  displayedKaruta,
  onClickRestart,
  onClickBack,
  onClickResult,
  onCloseDialog,
}: PresenterProps) => (
  <KarutaPlayingResult answers={answers} onClickBack={onClickBack} onClickRestart={onClickRestart}>
    <StyledBlock>
      {questions.map((q, i) => (
        <StyledQuestionJudgement
          key={q.id}
          data-test={`question-${q.id}`}
          karuta={q.correctKaruta}
          correct={answers[i].correct}
          onClick={onClickResult}
        />
      ))}
    </StyledBlock>
    <ClosableDialog
      open={!!displayedKaruta}
      onClose={onCloseDialog}
      PaperProps={{
        style: { maxWidth: 380 },
      }}
    >
      {displayedKaruta && <Material karuta={displayedKaruta} />}
    </ClosableDialog>
  </KarutaPlayingResult>
);

export const ExamResultContainer = ({ presenter, questions, answers, onClickRestart, onClickBack }: ContainerProps) => {
  const [displayedKaruta, openDialog, closeDialog] = useMaterialDialog();
  return presenter({
    questions,
    answers,
    displayedKaruta,
    onClickRestart,
    onClickBack,
    onClickResult: openDialog,
    onCloseDialog: closeDialog,
  });
};

export const ExamResult = (props: Props) => <ExamResultContainer presenter={ExamResultPresenter} {...props} />;

export const mapStateToProps = ({ questions }: GlobalState): ConnectedProps => ({
  questions: questions.questions ? questions.questions : [],
  answers: questions.answers ? questions.answers : [],
});

export default connect(mapStateToProps)(ExamResult);
