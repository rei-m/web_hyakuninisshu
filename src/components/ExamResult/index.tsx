import * as React from 'react';
import { Link } from 'react-router-dom';
import { withState } from 'recompose';
import styled from 'styled-components';
import { Dialog } from '@blueprintjs/core';
import Tatami from '../Tatami';
import QuestionResultsSummary from '../QuestionResultsSummary';
import QuestionResultsMap from '../QuestionResultsMap';
import KarutaCard from '../KarutaCard';
import { Answer, Karuta, Question } from '../../types';

export interface ExamResultProps {
  readonly totalCount: number;
  readonly correctCount: number;
  readonly averageAnswerSecond: number;
  readonly questions: Question[];
  readonly answers: Answer[];
  readonly onClickRestart: () => void;
}

const RootSection = Tatami.extend`
  width: 100vw;
  min-height: calc(100vh - 56px);

  @media screen and (min-width: 768px) {
    min-height: calc(100vh - 64px);
  }
`;

const Inner = styled.div`
  max-width: 380px;
  padding: 16px;
  margin: auto;
`;

const Button = styled.button`
  margin: 16px;
`;

const enhance = withState<
  {},
  Karuta | undefined,
  'displayedKaruta',
  'setDisplayedKaruta'
>('displayedKaruta', 'setDisplayedKaruta', undefined);

const ExamResult = enhance<
  ExamResultProps & {
    displayedKaruta: Karuta | undefined;
    setDisplayedKaruta: (karuta?: Karuta) => Karuta | undefined;
  }
>(
  ({
    averageAnswerSecond,
    correctCount,
    totalCount,
    questions,
    answers,
    onClickRestart,
    displayedKaruta,
    setDisplayedKaruta
  }) => {
    const onClickResultsMap = (karuta: Karuta) => {
      setDisplayedKaruta(karuta);
    };

    const onCloseDialog = () => {
      setDisplayedKaruta(undefined);
    };

    return (
      <RootSection>
        <Inner>
          <QuestionResultsSummary
            title={'正解数'}
            value={`${correctCount} / ${totalCount}`}
            style={{ marginBottom: 16 }}
          />
          <QuestionResultsSummary
            title={'平均回答時間'}
            value={`${averageAnswerSecond}秒`}
            style={{ marginBottom: 16 }}
          />
          <QuestionResultsMap
            questions={questions}
            answers={answers}
            onClickResult={onClickResultsMap}
            style={{ marginBottom: 16 }}
          />
          {correctCount !== totalCount && (
            <Button
              onClick={onClickRestart}
              className="pt-button pt-large pt-icon-repeat"
            >
              間違えた歌の練習をする
            </Button>
          )}
          <Link to="/exam" replace={true}>
            <Button
              className="pt-button pt-large pt-icon-undo"
              style={{ marginTop: 0 }}
            >
              メニューに戻る
            </Button>
          </Link>
        </Inner>
        <Dialog
          isOpen={!!displayedKaruta}
          onClose={onCloseDialog}
          title="正解"
          style={{
            maxWidth: 380,
            padding: 0,
            width: '80vw'
          }}
        >
          {displayedKaruta && <KarutaCard karuta={displayedKaruta} />}
        </Dialog>
      </RootSection>
    );
  }
);

export default ExamResult;
