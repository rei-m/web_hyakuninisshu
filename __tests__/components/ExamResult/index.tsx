import * as ReactTestRenderer from 'react-test-renderer';
import * as React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { shallow } from 'enzyme';
import { create } from '../../factories';
import { sel } from '../../helpers';
import ExamResult, { ExamResultProps } from '@src/components/ExamResult';
import QuestionResultsMap from '@src/components/QuestionResultsMap';
import { Answer, Question } from '@src/types';

describe('<ExamResult />', () => {
  const questions = [
    create<Question>('question'),
    create<Question>('question'),
    create<Question>('question')
  ];

  const answers = [
    create<Answer>('answer'),
    create<Answer>('answer'),
    create<Answer>('answer')
  ];

  let baseProps: ExamResultProps;

  beforeEach(() => {
    baseProps = {
      answers,
      averageAnswerSecond: 3.5,
      correctCount: 2,
      onClickRestart: jest.fn(),
      questions,
      totalCount: 3
    };
  });

  it('should render component when correct all', () => {
    const props = { ...baseProps, correctCount: 3 };
    const renderer = ReactTestRenderer.create(
      <MemoryRouter>
        <ExamResult {...props} />
      </MemoryRouter>
    );
    expect(renderer.toJSON()).toMatchSnapshot();
  });

  it('should render component when wrong', () => {
    const renderer = ReactTestRenderer.create(
      <MemoryRouter>
        <ExamResult {...baseProps} />
      </MemoryRouter>
    );
    expect(renderer.toJSON()).toMatchSnapshot();
  });

  it('should fire onClickRestart when restart clicked', () => {
    const wrapper = shallow(<ExamResult {...baseProps} />).dive();
    wrapper.find(sel('restart-training')).simulate('click');
    expect(baseProps.onClickRestart).toHaveBeenCalled();
  });

  it('should set state when QuestionResult clicked', () => {
    const wrapper = shallow(<ExamResult {...baseProps} />);
    wrapper
      .dive()
      .find(QuestionResultsMap)
      .props()
      .onClickResult(questions[0].correctKaruta);
    expect(wrapper.state().stateValue).toEqual(questions[0].correctKaruta);
  });
});
