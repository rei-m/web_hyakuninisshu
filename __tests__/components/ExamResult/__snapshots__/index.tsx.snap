// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExamResult /> should render component when correct all 1`] = `
<div
  className="sc-jTzLTM dfLKYC"
>
  <div
    className="sc-fjdhpX gcrVJE"
  >
    <div
      className="sc-bwzfXH lOjaS"
      style={
        Object {
          "marginBottom": 16,
        }
      }
    >
      <div
        className="sc-htpNat fCqHKh"
      >
        正解数
      </div>
      <div
        className="sc-bxivhb dTntfs"
      >
        3 / 3
      </div>
    </div>
    <div
      className="sc-bwzfXH lOjaS"
      style={
        Object {
          "marginBottom": 16,
        }
      }
    >
      <div
        className="sc-htpNat fCqHKh"
      >
        平均回答時間
      </div>
      <div
        className="sc-bxivhb dTntfs"
      >
        3.5秒
      </div>
    </div>
    <div
      className="sc-ifAKCX jmhkWb"
      style={
        Object {
          "marginBottom": 16,
        }
      }
    >
      <div
        className="sc-EHOje eapnbc"
        data-test="question-1"
        onClick={[Function]}
      >
        <span
          className="sc-bZQynM eIYJdI"
        >
          一番
        </span>
        <span
          className="sc-gzVnrw MJqTB"
        >
          <img
            src="check_correct.png"
            style={
              Object {
                "width": "80%",
              }
            }
          />
        </span>
      </div>
      <div
        className="sc-EHOje eapnbc"
        data-test="question-2"
        onClick={[Function]}
      >
        <span
          className="sc-bZQynM eIYJdI"
        >
          一番
        </span>
        <span
          className="sc-gzVnrw MJqTB"
        >
          <img
            src="check_correct.png"
            style={
              Object {
                "width": "80%",
              }
            }
          />
        </span>
      </div>
      <div
        className="sc-EHOje eapnbc"
        data-test="question-3"
        onClick={[Function]}
      >
        <span
          className="sc-bZQynM eIYJdI"
        >
          一番
        </span>
        <span
          className="sc-gzVnrw MJqTB"
        >
          <img
            src="check_correct.png"
            style={
              Object {
                "width": "80%",
              }
            }
          />
        </span>
      </div>
    </div>
    <a
      href="/exam"
      onClick={[Function]}
    >
      <button
        className="pt-button pt-large pt-icon-undo sc-jzJRlG hkqcLA"
        style={
          Object {
            "marginTop": 0,
          }
        }
      >
        メニューに戻る
      </button>
    </a>
  </div>
</div>
`;

exports[`<ExamResult /> should render component when wrong 1`] = `
<div
  className="sc-jTzLTM dfLKYC"
>
  <div
    className="sc-fjdhpX gcrVJE"
  >
    <div
      className="sc-bwzfXH lOjaS"
      style={
        Object {
          "marginBottom": 16,
        }
      }
    >
      <div
        className="sc-htpNat fCqHKh"
      >
        正解数
      </div>
      <div
        className="sc-bxivhb dTntfs"
      >
        2 / 3
      </div>
    </div>
    <div
      className="sc-bwzfXH lOjaS"
      style={
        Object {
          "marginBottom": 16,
        }
      }
    >
      <div
        className="sc-htpNat fCqHKh"
      >
        平均回答時間
      </div>
      <div
        className="sc-bxivhb dTntfs"
      >
        3.5秒
      </div>
    </div>
    <div
      className="sc-ifAKCX jmhkWb"
      style={
        Object {
          "marginBottom": 16,
        }
      }
    >
      <div
        className="sc-EHOje eapnbc"
        data-test="question-1"
        onClick={[Function]}
      >
        <span
          className="sc-bZQynM eIYJdI"
        >
          一番
        </span>
        <span
          className="sc-gzVnrw MJqTB"
        >
          <img
            src="check_correct.png"
            style={
              Object {
                "width": "80%",
              }
            }
          />
        </span>
      </div>
      <div
        className="sc-EHOje eapnbc"
        data-test="question-2"
        onClick={[Function]}
      >
        <span
          className="sc-bZQynM eIYJdI"
        >
          一番
        </span>
        <span
          className="sc-gzVnrw MJqTB"
        >
          <img
            src="check_correct.png"
            style={
              Object {
                "width": "80%",
              }
            }
          />
        </span>
      </div>
      <div
        className="sc-EHOje eapnbc"
        data-test="question-3"
        onClick={[Function]}
      >
        <span
          className="sc-bZQynM eIYJdI"
        >
          一番
        </span>
        <span
          className="sc-gzVnrw MJqTB"
        >
          <img
            src="check_correct.png"
            style={
              Object {
                "width": "80%",
              }
            }
          />
        </span>
      </div>
    </div>
    <button
      className="pt-button pt-large pt-icon-repeat sc-jzJRlG hkqcLA"
      data-test="restart-training"
      onClick={[MockFunction]}
    >
      間違えた歌の練習をする
    </button>
    <a
      href="/exam"
      onClick={[Function]}
    >
      <button
        className="pt-button pt-large pt-icon-undo sc-jzJRlG hkqcLA"
        style={
          Object {
            "marginTop": 0,
          }
        }
      >
        メニューに戻る
      </button>
    </a>
  </div>
</div>
`;
